test/cctest/test_environment.cc

```diff
@@ -32,24 +32,6 @@ class EnvironmentTest : public EnvironmentTestFixture {
   }
 };
 
-TEST_F(EnvironmentTest, PreExeuctionPreparation) {
-  const v8::HandleScope handle_scope(isolate_);
-  const Argv argv;
-  Env env {handle_scope, argv};
-
-  v8::Local<v8::Context> context = isolate_->GetCurrentContext();
-
-  const char* run_script = "process.argv0";
```

Comment:
The execution of `environment.js` was added to keep this test passing, which expected things like `process.argv0` to be set during `CreateEnvironment`. So I guess the key is: should the pre-execution steps (which sets up e.g. `process.argv0` here) be run in `CreateEnvironment()`, or `LoadEnvironment()`?

My original thought before finding the test broken had been that it should be run in `LoadEnvironment()`, as is done in all the non-embedding entry points (through explicitly running functions exported through `require('internal/bootstrap/pre_execution')`), but it would be a breaking change (which is exactly what this PR is doing and why it needs to remove this test).

I am good with finally breaking this somewhat odd contract but I wonder if anyone relies on it? Also it probably would be nicer if we keep `environment.js` and add a method that runs it, then we could just run that method here to set up `process.argv0` and do not need to remove this test. Having control over when the pre-execution are run from C++ without writing `require('internal/bootstrap/pre_execution')` in JS themselves sounds useful to an embedder.

Reply:
I'm not aware of anyone relying on this behavior, but I think I would be surprised to determine that anyone did. It seems to me that any points of divergence between `CreateEnvironment` exposed externally and `CreateMainEnvironment` used internally will create behavioral patterns we don't want, and since Electron seeks to mimic Node.js behavior when being run in `ELECTRON_RUN_AS_NODE` mode as much as possible i'd prefer to prevent them as much as possible.

I do understand the desire to avoid breaking the test; I'm curious though with regard to 

> Having control over when the pre-execution are run from C++ without writing require('internal/bootstrap/pre_execution') in JS themselves sounds useful to an embedder.

what use cases you see this solving? i'm happy to potentially rework this commit to keep `environment` but with regard to my earlier comment i personally believe we should avoid `CreateEnvironment` doing anything that `CreateMainEnvironment` doesn't also do.

Reply:
@codebytere I see, thanks for the explanation, I was thinking about that the combination of `CreateEnvironment()` and `LoadEnvironment()` do not allow an embedder to run Node the way Electron usually runs without patching (which was along the lines with the comments in `LoadEnvironment()` that these CLI option switches may not make sense for all embedders). But the use case of `ELECTRON_RUN_AS_NODE` is different and where the switches are necessary.

> what use cases you see this solving

For example, `CreateMainEnvironment()` creates an Environment suitable only for main thread execution, so this naturally rules out the possibility to create multiple Environments from the same thread. Keeping the pre-execution initialization steps separate makes it possible to fixup and/or snapshot the Environment before the creation of things like IPC channels, especially allowing them to perform the fixup through JS scripts making use of the internal `require` (which needs to be setup in `CreateEnvironment()`), which seems potentially useful for bundlers. A use case I can imagine is `CreateEnvironment()` to setup a minimal `Environment`, fix up the Environment (e.g. injecting the embedder's own builtins), and creating a snapshot out of it to be distributed around. Then in the distributed builds, `CreateEnvironment()` from a snapshot, perform some fixups, and `InitializeEnvironment()` for pre-execution steps before `LoadEnvironment()`. I am imagining the fixups could for instance, involve hooking fs up in the JS land with a virtual file system (since the first actual invocation of `fs` would be in pre-execution to pre-load user modules).

Reply:
@joyeecheung i do see that and would be interested in working on addressing that - would you be amenable to potentially addressing that use case in a follow up and merging this as-is?

Reply:
I am good with the behavior change itself, but do note that if this land as-is, the removed code would need to be added back in the follow-up instead of just being modified.

Reply:
Yeah, I’ll be rebasing #30467 against this when this lands – it should be easy enough to keep the test working then.

Reply:
that works! i'll address this as soon :)